### PR TITLE
Distance table AA type offload

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
Initial version. Correctness guaranteed. More performance tuning and adjust API calls will be needed.

## What type(s) of changes does this code introduce?
- New feature
- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
epyc-server

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

**Comment by ye-luo** [comment]
Test this please

**Review by prckent — COMMENTED** [review]
Confirming that I could build and run this PR.

What are your current plans for subsequent PRs to complete plumbing this in and activating it? There are some places where a few more comments would be ideal (function definitions etc.) but if this is more exploratory code and likely to evolve much further, I think those updates can wait so that you can keep developing.

**Comment by ye-luo** [comment]
@prckent Please leave review comments about missing documentation and additional helpful comments. The I can decide to add them right now preferably or later if an API will be changed.

File: src/Particle/DynamicCoordinates.h [hunk 1]
```diff
@@ -55,6 +56,14 @@ class DynamicCoordinates
 
   virtual void setAllParticlePos(const ParticlePos_t& R)         = 0;
   virtual void setOneParticlePos(const PosType& pos, size_t iat) = 0;
+  virtual void mw_copyActiveOldParticlePos(const RefVector<DynamicCoordinates>& coords_list,
```
Comment (prckent) [inline src/Particle/DynamicCoordinates.h:59]: Definitions are needed here, and broadly for all the functions in DynamicCoordinates

File: src/Particle/DynamicCoordinates.h [hunk 2]
```diff
@@ -55,6 +56,14 @@ class DynamicCoordinates
...
+                                           size_t iat,
+                                           const std::vector<PosType>& new_positions) {}
+
+  virtual void mw_acceptParticlePos(const RefVector<DynamicCoordinates>& coords_list,
```
Comment (prckent) [inline src/Particle/DynamicCoordinates.h:63]: Ditto. Definition.

File: src/Particle/Lattice/ParticleBConds3DSoa.h [hunk 3]
```diff
@@ -449,40 +433,37 @@ struct DTD_BConds<T, 3, PPPG + SOA_OFFSET>
...
+    const T dely = displ_1 + ar_0 * r01 + ar_1 * r11 + ar_2 * r21;
+    const T delz = displ_2 + ar_0 * r02 + ar_1 * r12 + ar_2 * r22;
 
-      temp_r[iat] = std::sqrt(rmin);
-      dx[iat]     = flip * (delx + cellx[ic]);
-      dy[iat]     = flip * (dely + celly[ic]);
-      dz[iat]     = flip * (delz + cellz[ic]);
+    T rmin = delx * delx + dely * dely + delz * delz;
+    int ic = 0;
+#pragma unroll(7)
```
Comment (prckent) [inline src/Particle/Lattice/ParticleBConds3DSoa.h:452]: Is this pragma even needed these days? Was it ever needed? Compiler can see the loop size and take the appropriate decision for the architecture.

File: src/Particle/Lattice/ParticleBConds3DSoa.h [hunk 4]
```diff
@@ -608,37 +588,34 @@ struct DTD_BConds<T, 3, PPNG + SOA_OFFSET>
...
+    const T delx = displ_0 + ar_0 * r00 + ar_1 * r10;
+    const T dely = displ_1 + ar_0 * r01 + ar_1 * r11;
 
-      temp_r[iat] = std::sqrt(rmin + delz * delz);
-      dx[iat]     = flip * (delx + cellx[ic]);
-      dy[iat]     = flip * (dely + celly[ic]);
-      dz[iat]     = delz;
+    T rmin = delx * delx + dely * dely;
+    int ic = 0;
+#pragma unroll(3)
```
Comment (prckent) [inline src/Particle/Lattice/ParticleBConds3DSoa.h:605]: Ditto previous pragma comment.

File: src/Particle/ParticleSet.cpp [hunk 5]
```diff
@@ -497,30 +500,36 @@ void ParticleSet::computeNewPosDistTablesAndSK(Index_t iat, const SingleParticle
...
-      for (int iw = 0; iw < P_list.size(); iw++)
-        P_list[iw].get().SK->makeMove(iat, new_positions[iw]);
+      const auto dt_list(extractDTRefList(p_list, i));
+      leader.DistTables[i]->mw_move(dt_list, p_list, new_positions, iat, maybe_accept);
     }
   }
+  auto& SK = leader.SK;
+  if (SK && SK->DoUpdate)
+  {
+#pragma omp parallel for
```
Comment (prckent) [inline src/Particle/ParticleSet.cpp:529]: Eventually need an SK mw_ call here to hide the omp here, as with the mw_move call above.

File: src/Particle/RealSpacePositions.h [hunk 6]
```diff
@@ -43,6 +43,16 @@ class RealSpacePositions : public DynamicCoordinates
   }
   void setOneParticlePos(const PosType& pos, size_t iat) override { RSoA(iat) = pos; }
 
+  void mw_acceptParticlePos(const RefVector<DynamicCoordinates>& coords_list,
+                            size_t iat,
+                            const std::vector<PosType>& new_positions,
+                            const std::vector<bool>& isAccepted) override
+  {
+    for (size_t iw = 0; iw < isAccepted.size(); iw++)
+      if (isAccepted[iw])
```
Comment (prckent) [inline src/Particle/RealSpacePositions.h:52]: If setoneparticle pos is ever not cheap, would need an omp loop here. Is set() guaranteed to be cheap?

File: src/Particle/RealSpacePositionsOMPTarget.h [hunk 7]
```diff
@@ -60,11 +59,13 @@ class RealSpacePositionsOMPTarget : public DynamicCoordinates
     resize(R.size());
     RSoA_hostview.copyIn(R);
     updateH2D();
+    is_nw_new_pos_prepared = false;
```
Comment (prckent) [inline src/Particle/RealSpacePositionsOMPTarget.h:62]: Somewhere should describe what having a position prepared means.

**Review by prckent — APPROVED** [review]

File: src/Particle/Lattice/ParticleBConds3DSoa.h (hunk 3 above)
Comment (ye-luo) [reply src/Particle/Lattice/ParticleBConds3DSoa.h:452]: Not sure. So should be checked when we really look at performance.

File: src/Particle/ParticleSet.cpp (hunk 5 above)
Comment (ye-luo) [reply src/Particle/ParticleSet.cpp:529]: It is still unclear tome  which feature needs updating SK during pbyp. My best hope is delete.

File: src/Particle/RealSpacePositions.h (hunk 6 above)
Comment (ye-luo) [reply src/Particle/RealSpacePositions.h:52]: Still exploring... for now it is very cheap.

**Comment by ye-luo** [comment]
I prefer merge and I will deal with the documentation on my very soon PR.

**Pull request merged**